habitat_baselines/il/models/models.py

```diff
@@ -0,0 +1,780 @@
+            nn.BatchNorm2d(512),
+            nn.ReLU(inplace=True),
+            nn.Dropout2d(),
+            nn.Conv2d(512, 512, 1),
+            nn.BatchNorm2d(512),
+            nn.ReLU(inplace=True),
+            nn.Dropout2d(),
+        )
+
+        self.encoder_seg = nn.Conv2d(512, self.num_classes, 1)
+        self.encoder_depth = nn.Conv2d(512, 1, 1)
+        self.encoder_ae = nn.Conv2d(512, 3, 1)
+
+        self.score_pool2_seg = nn.Conv2d(16, self.num_classes, 1)
+        self.score_pool3_seg = nn.Conv2d(32, self.num_classes, 1)
+
+        self.score_pool2_depth = nn.Conv2d(16, 1, 1)
+        self.score_pool3_depth = nn.Conv2d(32, 1, 1)
+
+        self.score_pool2_ae = nn.Conv2d(16, 3, 1)
+        self.score_pool3_ae = nn.Conv2d(32, 3, 1)
+
+        self.pretrained = pretrained
+
+        if self.pretrained is True:
+            print("Loading CNN weights from %s" % checkpoint_path)
+            checkpoint = torch.load(
+                checkpoint_path, map_location={"cuda:0": "cpu"}
+            )
+            self.load_state_dict(checkpoint)
+            for param in self.parameters():
+                param.requires_grad = False
+        else:
+            for m in self.modules():
+                if isinstance(m, nn.Conv2d):
+                    n = (
+                        m.kernel_size[0]
+                        * m.kernel_size[1]
+                        * (m.out_channels + m.in_channels)
+                    )
+                    m.weight.data.normal_(0, math.sqrt(2.0 / n))
+                elif isinstance(m, nn.BatchNorm2d):
+                    m.weight.data.fill_(1)
+                    m.bias.data.zero_()
+
+    def forward(self, x):
+        # assert self.training is False
+        conv1 = self.conv_block1(x)
+        conv2 = self.conv_block2(conv1)
+        conv3 = self.conv_block3(conv2)
+        conv4 = self.conv_block4(conv3)
+        return conv4.view(-1, 32 * 12 * 12)
+
+
+class QuestionLstmEncoder(nn.Module):
+    def __init__(
+        self,
+        token_to_idx,
+        wordvec_dim=64,
+        rnn_dim=64,
+        rnn_num_layers=2,
+        rnn_dropout=0,
+    ):
+        super(QuestionLstmEncoder, self).__init__()
+
+        self.token_to_idx = token_to_idx
+        self.NULL = token_to_idx["<pad>"]
+        self.START = token_to_idx["<s>"]
+        self.END = token_to_idx["</s>"]
+
+        self.embed = nn.Embedding(len(token_to_idx), wordvec_dim)
+        self.rnn = nn.LSTM(
+            wordvec_dim,
+            rnn_dim,
+            rnn_num_layers,
+            dropout=rnn_dropout,
+            batch_first=True,
+        )
+
+        self.init_weights()
+
+    def init_weights(self):
+        initrange = 0.1
+        self.embed.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, x):
+        N, T = x.size()
+        idx = torch.LongTensor(N).fill_(T - 1)
+
+        # Find the last non-null element in each sequence
+        x_cpu = x.data.cpu()
+        for i in range(N):
```

Comment:
No for loop, this will be unnecessarily slow.  You can just do `idx = (x != self.NULL).long.sum(-1) - 1`